System/IO/BlockIO.hs

```diff
@@ -235,24 +236,24 @@ submitIO IOCtx {
       Nothing -> throwIO closed
       Just uring -> do
 --      print ("submitIO", iobatchOpCount)
-        sequence_
-          [ --print ioop >>
-            case ioop of
-            IOOpRead  fd off buf bufOff cnt -> do
-              guardPinned buf
-              URing.prepareRead  uring fd off
-                                 (mutableByteArrayContents buf `plusPtr` bufOff)
-                                 cnt ioopid
-            IOOpWrite fd off buf bufOff cnt -> do
-              guardPinned buf
-              URing.prepareWrite uring fd off
-                                 (mutableByteArrayContents buf `plusPtr` bufOff)
-                                 cnt ioopid
-          | (ioop, ioopix) <- zip ioops [IOOpIx 0 ..]
-          , let !ioopid = packIOOpId iobatchIx ioopix ]
+        V.iforM_ ioops $ \ioopix ioop ->
+            let !ioopid = packIOOpId iobatchIx (IOOpIx $ fromIntegral ioopix)
+            in
+              --print ioop >>
+              case ioop of
+              IOOpRead  fd off buf bufOff cnt -> do
+                guardPinned buf
+                URing.prepareRead  uring fd off
+                                  (mutableByteArrayContents buf `plusPtr` bufOff)
+                                  cnt ioopid
+              IOOpWrite fd off buf bufOff cnt -> do
+                guardPinned buf
+                URing.prepareWrite uring fd off
+                                  (mutableByteArrayContents buf `plusPtr` bufOff)
+                                  cnt ioopid
         URing.submitIO uring
 --      print ("submitIO", "submitting done")
-    map (IOResult_ . coerce) . elems <$> takeMVar iobatchCompletion
+    fmap (IOResult_ . coerce) . V.fromList . elems <$> takeMVar iobatchCompletion
```

Comment:
The `V.fromList` here is not ideal. Maybe we should use `Vector`s internally in the completion thread @dcoutts?

Reply:
Yes I agree. Done!